tests/symlinked_included_files.test.ts

```diff
@@ -0,0 +1,65 @@
+import { readdir } from 'fs/promises'
+import { join } from 'path'
+
+import decompress from 'decompress'
+import { dir as getTmpDir } from 'tmp-promise'
+import { expect, test } from 'vitest'
+
+import { ARCHIVE_FORMAT, zipFunction } from '../src/main.js'
+
+import { FIXTURES_ESM_DIR } from './helpers/main.js'
+
+const readDirWithTypes = async (dir: string, readFiles?: Record<string, boolean>, parent = '') => {
+  const files: Record<string, boolean> = readFiles || {}
+  const dirents = await readdir(dir, { withFileTypes: true })
+
+  for (const dirent of dirents) {
+    if (dirent.isDirectory()) {
+      await readDirWithTypes(join(dir, dirent.name), files, dirent.name)
+    } else {
+      files[join(parent, dirent.name)] = dirent.isSymbolicLink()
+    }
+  }
+
+  return files
+}
+
+test('if symlinked directories from the included files are preserved', async () => {
+  const { path: tmpDir } = await getTmpDir({ prefix: 'zip-it-test' })
+  const basePath = join(FIXTURES_ESM_DIR, 'symlinked-included-files')
+  const mainFile = join(basePath, 'function.mjs')
+
+  // assert on the source files
+  expect(await readDirWithTypes(basePath)).toEqual({
+    'function.mjs': false,
+    'crazy-dep/package.json': false,
+    'node_modules/crazy-dep': true,
+  })
+
+  const result = await zipFunction(mainFile, tmpDir, {
+    archiveFormat: ARCHIVE_FORMAT.ZIP,
+    basePath,
+    config: {
+      '*': {
+        includedFiles: ['**'],
+      },
+    },
+    featureFlags: {},
+    repositoryRoot: basePath,
+    systemLog: console.log,
+    debug: true,
+    internalSrcFolder: undefined,
+  })
+
+  const unzippedPath = join(tmpDir, 'extracted')
+  await decompress(result!.path, unzippedPath)
```

Comment:
Does the [existing util for unzipping](https://github.com/netlify/zip-it-and-ship-it/blob/8a3845d6e1500eba8d07c1e29b1176d5c9ca7d85/tests/helpers/main.ts#L142) not work here due to symlinks? If that's the case, it would be good to migrate all tests to use this new library, otherwise we're left with two distinct ways of doing the same thing.

Not saying you need to do that as part of this PR, but at least let's have an issue for it.

Reply:
(Just saw [your comment](https://github.com/netlify/zip-it-and-ship-it/pull/1701#issuecomment-1912012374) about this.)